codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java

```diff
@@ -641,6 +647,30 @@ public void operationComplete(ChannelFuture future) throws Exception {
         assertTrue(listenerExecuted.get());
     }
 
+    @Test
+    public void upgradeEventNoRefCntError() throws Http2Exception {
+        frameListener.onHeadersRead(http2HandlerCtx, Http2CodecUtil.HTTP_UPGRADE_STREAM_ID, request, 31, false);
+
+        final FullHttpRequest request = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
+        final HttpServerUpgradeHandler.UpgradeEvent upgradeEvent = mock(HttpServerUpgradeHandler.UpgradeEvent.class);
+        when(upgradeEvent.retain()).thenAnswer(new Answer<HttpServerUpgradeHandler.UpgradeEvent>() {
+            @Override
+            public HttpServerUpgradeHandler.UpgradeEvent answer(InvocationOnMock invocationOnMock) throws Throwable {
+                request.retain();
+                return upgradeEvent;
+            }
+        });
+        when(upgradeEvent.release()).thenAnswer(new Answer<Boolean>() {
+            @Override
+            public Boolean answer(InvocationOnMock invocationOnMock) throws Throwable {
+                return request.release();
+            }
+        });
+
+        when(upgradeEvent.upgradeRequest()).thenReturn(request);
+        channel.pipeline().fireUserEventTriggered(upgradeEvent);
```

Comment:
can we also check the reference count on upgradeEvent directly on the request here?

Reply:
sure... that said without the fix it fails the test because it tries to release everything that is stored in the `EmbeddedChannel`, so I think its not really needed.

Reply:
yeah, I figured it might be nice to help avoid regressions where it's +1 instead of -1.

Reply:
good point let me do this

Reply:
done